python/cuml/cuml/experimental/accel/estimator_proxy.py

```diff
@@ -0,0 +1,146 @@
+#
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import cuml
+import importlib
+import inspect
+import os
+
+from cuml.internals.global_settings import GlobalSettings
+from cuml.internals.mem_type import MemoryType
+from cuml.internals import logger
+from cuml.internals.safe_imports import gpu_only_import, cpu_only_import
+from typing import Optional
+
+
+# currently we just use this dictionary for debugging purposes
+patched_classes = {}
+
+
+def intercept(
+    original_module: str,
+    accelerated_module: str,
+    original_class_name: str,
+    accelerated_class_name: Optional[str] = None,
+):
+
+    if accelerated_class_name is None:
+        accelerated_class_name = original_class_name
+    # Import the original host module and cuML
+    module_a = cpu_only_import(original_module)
+    module_b = gpu_only_import(accelerated_module)
+
+    # Store a reference to the original (CPU) class
+    if original_class_name in patched_classes:
+        original_class_a = patched_classes[original_class_name]
+    else:
+        original_class_a = getattr(module_a, original_class_name)
+        patched_classes[original_class_name] = original_class_a
+
+    # Get the class from cuML so ProxyEstimator inherits from it
+    class_b = getattr(module_b, accelerated_class_name)
+
+    # todo (dgd): add environment variable to disable this
+    class ProxyEstimatorMeta(cuml.internals.base_helpers.BaseMetaClass):
+        def __repr__(cls):
+            return repr(original_class_a)
+
+    class ProxyEstimator(class_b, metaclass=ProxyEstimatorMeta):
```

Comment:
What was the thinking around making this inherit from `class_b` (the cuml class right?)? Naively I'd have made this a class that doesn't inherit from either class A or class B, but has an instance of each stored as an attribute and then proxies to the appropriate one.

The next thing I'd have tried is inheriting from the sklearn class, in the hopes of making `isinstance` checks "just work". Though I think I'd probably gone back to using attributes because the asymmetry for proxying A and B would have confused me.

Looking at this class it mostly does hyper-parameter translation and then deals with serialisation. I think this is because the cuml class that we inherit from already has all the methods we need and they get called directly. Which might be the answer to my original question?

Reply:
The machinery of dispatching lives in the cuML class, otherwise we would need to recreate dispatching here when we already have in cuML.